### PR TITLE
vimPlugins: use lua derivation if it exists

### PR DESCRIPTION
**Pull request by teto**
Neovim plugins are now more often than not written in lua.
One advantage of the lua ecosystem over vim's is the existence of
luarocks and the rockspec format, which allows to specify a package
dependencies formally.
I would like more neovim plugins to have a formal description,
"rockspec" being the current candidate.
This MR allows to use nix lua packages as neovim plugins, so as to enjoy
every benefit that rockspecs bring:
- dependdency discovery
- ability to run test suite
- luarocks versioning
- rockspec metadata

the vim update.py script will check if an attribute with the vim plugin
pname exists in lua51Packages. If it does, it uses
buildNeovimPluginFrom2Nix on it, which modifies the luarocks config to
do an almost flat install (luarocks will install the package in the lua
folder instead of share/5.1/lua etc).
It also calls toVimPlugin on it to get all the vim plugin niceties.

The list of packages that could benefit from this is available at
https://luarocks.org/labels/neovim
but I hope it grows.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/interpreters/lua-5/hooks/luarocks-move-data.sh [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+luarocksMoveDataHook () {
+    echo "Executing luarocksMoveDataHook"
+    if [ -d "$out/$rocksSubdir" ]; then
+        cp -rfv "$out/$rocksSubdir/$pname/$version/." "$out"
+    fi
+
+    echo "Finished executing luarocksMoveDataHook"
+}
+
+echo "Using luarocksMoveDataHook"
```
Comment (SuperSandro2000) [inline pkgs/development/interpreters/lua-5/hooks/luarocks-move-data.sh:14]: Isn't that a bit much noise for just copying some files?

File: pkgs/applications/editors/vim/plugins/vim-utils.nix [hunk 2]
```diff
@@ -566,4 +560,21 @@ rec {
       nativePlugins = lib.concatMap ({start?[], opt?[], knownPlugins?vimPlugins}: start++opt) nativePluginsConfigs;
     in
       nativePlugins ++ nonNativePlugins;
+
+  toVimPlugin = drv:
+    drv.overrideAttrs(oldAttrs: {
+      # dont move the "doc" folder since vim expects it
+      forceShare= [ "man" "info" ];
```
Comment (SuperSandro2000) [inline pkgs/applications/editors/vim/plugins/vim-utils.nix:567]:
```suggestion
      forceShare = [ "man" "info" ];
```

File: pkgs/applications/editors/vim/plugins/default.nix [hunk 3]
```diff
@@ -9,23 +12,11 @@ let
   inherit (lib) extends;
 
   initialPackages = self: {
```
Comment (SuperSandro2000) [inline pkgs/applications/editors/vim/plugins/default.nix:14]:
```suggestion
```

File: pkgs/applications/editors/vim/plugins/default.nix [hunk 4]
```diff
@@ -9,23 +12,11 @@ let
...
-
-        nativeBuildInputs = oldAttrs.nativeBuildInputs or [] ++ [
-          vimGenDocHook
-          vimCommandCheckHook
-        ];
-        passthru = (oldAttrs.passthru or {}) // {
-          vimPlugin = true;
-        };
-      });
   };
```
Comment (SuperSandro2000) [inline pkgs/applications/editors/vim/plugins/default.nix:15]:
```suggestion
```

File: pkgs/applications/editors/neovim/build-neovim-plugin.nix [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+, buildVimPluginFrom2Nix
+, buildLuarocksPackage
+, lua51Packages
+, toVimPlugin
+}:
+let
+  # sanitizeDerivationName
+  normalizeName = name:
+    (lib.concatMapStrings (s: if lib.isList s then "-" else s))
+      ((builtins.split "[.]") name);
```
Comment (pennae) [inline pkgs/applications/editors/neovim/build-neovim-plugin.nix:12]:
```suggestion
  normalizeName = lib.replaceStrings [ "." ] [ "-" ];
```

File: pkgs/applications/editors/neovim/utils.nix [hunk 6]
```diff
@@ -184,4 +186,9 @@ in
 {
   inherit makeNeovimConfig;
   inherit legacyWrapper;
+
+  inherit (callPackage ./build-neovim-plugin.nix {
+    inherit (vimUtils) buildVimPluginFrom2Nix toVimPlugin;
+    inherit buildLuarocksPackage;
+  }) buildNeovimPluginFrom2Nix;
```
Comment (pennae) [inline pkgs/applications/editors/neovim/utils.nix:193]: while it's not a problem here this is generally not a great pattern. the `x` in `inherit (x) ys ...` is evaluated indepdently for every `y`, so adding something here would evaluate the whole thing many times

File: pkgs/applications/editors/vim/plugins/default.nix (hunk 4 above)
Comment (teto) [reply pkgs/applications/editors/vim/plugins/default.nix:15]: sry I dont get the issue here

File: pkgs/applications/editors/neovim/utils.nix (hunk 6 above)
Comment (teto) [reply pkgs/applications/editors/neovim/utils.nix]: ha I had no idea thanks, I've removed the outer "inherit".

File: pkgs/development/interpreters/lua-5/hooks/luarocks-move-data.sh (hunk 1 above)
Comment (teto) [reply pkgs/development/interpreters/lua-5/hooks/luarocks-move-data.sh:14]: this is one of the tricky part so I wanted it to stand out:
- there is no way to tell luarocks to install "data" in the root folder 
- it is easy to plug in only for vim plugins and makes it clear it only concerns these plugins

**Comment by thatsmydoing** [comment]
I think this causes a number of problems. For one, `plenary-nvim` is missing a `data` folder so filetype detection is broken. Also, this silently replaces the plugin definition with the `luarocks` version which does not necessarily match the `src` specified in `buildNeovimPluginFrom2Nix`. Case in point, `plenary-nvim` is now a year out of date since `luarocks` is not frequently updated.

I think by default, we just want vim plugins to mirror latest github as that's what other vim package managers do.

**Comment by teto** [comment]
> For one, plenary-nvim is missing a data folder so filetype detection is broken. Also, this silently replaces the plugin definition with the luarocks version which does not necessarily match the src specified in buildNeovimPluginFrom2Nix. Case in point, plenary-nvim is now a year out of date since luarocks is not frequently updated.

These are valid concerns, overriding the specified src silently as done right now is bad.

>I think by default, we just want vim plugins to mirror latest github as that's what other vim package managers do.

Nix can do better though. We can use the rockspec from the repo instead of the one from luarocks so it's not necessarily the issue.

I will coordinate with tjdevries to update the rockspec on the server as a quickfix. 
I have a few neovim-related changes in the pipeline. Once those are merged, I'll look into it.
Are you thinking about this folder ? it seems quite new and I wonder why it's in a "data" folder https://github.com/nvim-lua/plenary.nvim/tree/master/data/plenary/filetypes, I 'll look into it.

**Comment by thatsmydoing** [comment]
> Nix can do better though. We can use the rockspec from the repo instead of the one from luarocks so it's not necessarily the issue.

I think this would be better done once the luarocks ecosystem is a bit more mature and there is a critical mass of people actually using it to manage their nvim plugins outside of nix. Project authors themselves will also have to keep the rockspec up-to-date. I just feel it's too much too early.

> Are you thinking about this folder ? it seems quite new and I wonder why it's in a "data" folder https://github.com/nvim-lua/plenary.nvim/tree/master/data/plenary/filetypes, I 'll look into it.

It's from at least Dec 2020 so it was already there when the rockspec was added but maybe the spec is missing something for it?

**Comment by teto** [comment]
so I've opened https://github.com/NixOS/nixpkgs/issues/183304 to track the status of the telescope resolution.

As for the rockspec-based packages, you gotta start somewhere. There has been no complaint so far so I think it's overall ok with the issue that you have mentioned that can be improved. I've opened https://github.com/NixOS/nixpkgs/issues/185397 to track it. There is an effort to have neovim plugin format https://github.com/nvim-lua/nvim-package-specification

**Comment by figsoda** [comment]
Shouldn't this use luajit instead of lua 5.1?

**Pull request merged**